gilbreth_perception/src/recognition_node.cpp

```diff
@@ -0,0 +1,295 @@
+		fpfh_est.compute (*model_features);
+		model_features_list.push_back(model_features);
+	}
+
+
+	//ICP
+	std::vector<pcl::PointIndices>  recognized_indices;
+	pcl::PointCloud<PointType>::Ptr target (new pcl::PointCloud<PointType> ());
+	pcl::PointCloud<PointType>::Ptr recognized_cloud (new pcl::PointCloud<PointType> ());
+	pcl::PointCloud<pcl::FPFHSignature33>::Ptr target_features (new pcl::PointCloud<pcl::FPFHSignature33> );
+	pcl::SampleConsensusInitialAlignment<pcl::PointXYZ, pcl::PointXYZ, pcl::FPFHSignature33> sac_ia_;
+	std::vector<Result, Eigen::aligned_allocator<Result> > results_temp;
+	Result result;
+
+	// Intialize the parameters in the Sample Consensus Intial Alignment (SAC-IA) algorithm
+	sac_ia_.setMinSampleDistance (min_sample_distance_);
+	sac_ia_.setMaxCorrespondenceDistance (max_correspondence_distance_);
+	sac_ia_.setMaximumIterations (nr_iterations_);
+
+	int min_index;
+
+	sac_ia_.setInputTarget (scene);
+	sac_ia_.setTargetFeatures (scene_features);
+	results_temp.resize (model_list.size ());
+	float best_score=1;
+	for (int j=0;j<model_list.size();j++){
+		sac_ia_.setInputSource (model_list[j]);
+		sac_ia_.setSourceFeatures (model_features_list[j]);
+		pcl::PointCloud<pcl::PointXYZ> registration_output;
+		sac_ia_.align (registration_output);
+		results_temp[j].item_name=model_name[j];
+		results_temp[j].item_id=j;
+		results_temp[j].fitness_score = (float) sac_ia_.getFitnessScore (max_correspondence_distance_);
+		results_temp[j].final_transformation = sac_ia_.getFinalTransformation ();
+		std::cerr<<"model "<<j<<" FitnessScore "<<results_temp[j].fitness_score<<std::endl;
+
+		if (results_temp[j].fitness_score < best_score){
+			min_index=j;
+			best_score=results_temp[j].fitness_score;
+
+		}
+
+	}
+	result=results_temp[min_index];
+	std::cerr<<"Item_Name: "<<result.item_name<<std::endl;
+
+	pcl::PointCloud<PointType>::Ptr rotated_model(new pcl::PointCloud<PointType>());
+	pcl::transformPointCloud(*model_list[result.item_id], *rotated_model,result.final_transformation);
+
+
+
+
+	//  Visualize recognition result
+	viewer.removeAllPointClouds();
+	pcl::visualization::PointCloudColorHandlerCustom<PointType> scene_color_handler (scene, 137, 137, 137);
+	viewer.addPointCloud (scene,scene_color_handler, "scene_cloud");
+
+	pcl::visualization::PointCloudColorHandlerCustom<PointType> rotated_model_color_handler (rotated_model, 255, 0, 0);
+	viewer.addPointCloud (rotated_model, rotated_model_color_handler, "rotated_model");
+
+	pcl::PointCloud<PointType>::Ptr pick_points(new pcl::PointCloud<PointType>());
+	pick_points->push_back(rotated_model->points[pick_point_indices[result.item_id]]);
+	pcl::visualization::PointCloudColorHandlerCustom<PointType> pick_points_color_handler (pick_points, 0, 0, 255);
+	viewer.addPointCloud (pick_points, pick_points_color_handler, "pick_points");
+	viewer.setPointCloudRenderingProperties (pcl::visualization::PCL_VISUALIZER_POINT_SIZE, 10, "pick_points");
+
+	viewer.spinOnce ();
```

Comment:
We may make the viewer optional or preferably publish the point cloud to rviz for visualization purposes.

Reply:
Done, a switch will control it in yaml file, default is off.